doc/md: improving privacy docs and multi tenant example (#1724)
ent/ent: doc/md/privacy.md

	if _, err := client.Tenant.Create().Save(viewOnly); !errors.Is(err, privacy.Deny) {
		return fmt.Errorf("expect operation to fail, but got %w", err)
	}
	// Apply the same operation with "Admin" role, expect it to pass.

> 	adminCtx := viewer.NewContext(ctx, viewer.UserViewer{Role: viewer.Admin})
> 	hub, err := client.Tenant.Create().SetName("GitHub").Save(adminCtx)

Reviewer: ```suggestion
	admin := viewer.NewContext(ctx, viewer.UserViewer{Role: viewer.Admin})
	hub, err := client.Tenant.Create().SetName("GitHub").Save(admin)
```

Leave it as before. No need to add the type to the var name.

```

Now, we want to enforce that viewers can see only groups and users that are connected to the tenant they belong to.
In this case, we create another type of privacy rule named `FilterTenantRule`.  

> This rule can help us filter out entities that are not connected to the same tenant.

Reviewer: A `FilterRule` rule is a **type** of privacy rule. The `FilterTenantRule` is an implementation. 

I hope this gives more context, and I'll be happy to improve this paragraph if it's not clear, but please mention the `FilterRule` type.

Author: @a8m check it now, please.

> This rule can help us to filters out entities that are not connected to the same tenant.

Reviewer 2: ```suggestion
This rule can help us filter out entities that are not connected to the same tenant.
```

> > Note, the privacy filtering option needs to be enabled using the [`entql`](https://entgo.io/docs/feature-flags/#entql-filtering) feature-flag (see instructions [above](#configuration)).

Reviewer: ```suggestion
> Note, the privacy filtering option needs to be enabled using the [`entql`](features.md#entql-filtering) feature-flag (see instructions [above](#configuration)).
```

Use relative paths.

> > Note, the filtering option for privacy needs to be enabled using the [`entql`](https://entgo.io/docs/feature-flags/#entql-filtering) feature-flag (see instructions [above](#configuration)).

Reviewer 2: ```suggestion
> Note, the privacy filtering option needs to be enabled using the [`entql`](https://entgo.io/docs/feature-flags/#entql-filtering) feature-flag (see instructions [above](#configuration)).
```

```go title="examples/privacytenant/rule/rule.go"
// FilterTenantRule is a query/mutation rule that filters out entities that are not in the tenant.
func FilterTenantRule() privacy.QueryMutationRule {

> 	// TenantsFilter is an interface to wrap WhereHasTenantWith() predicate that's
> 	// used by both `Group` and `User` schemas.
> 	type TenantsFilter interface {

Reviewer: ```suggestion
	// TenantsFilter is an interface to wrap WhereHasTenantWith()
	// predicate that's used by both `Group` and `User` schemas.
	type TenantsFilter interface {
```

👍

Then, after running the code-generation, we expect the privacy-rules to take effect on the client operations.

```go title="examples/privacytenant/example_test.go"
func Do(ctx context.Context, client *ent.Client) error {
    // A continuation of the code-block above.

	// Create 2 users connected to the 2 tenants we created above

> 	a8mHub := client.User.Create().SetName("a8mHub").SetTenant(hub).SaveX(adminCtx)
> 	natiLab := client.User.Create().SetName("natiLab").SetTenant(lab).SetFoods([]string{"Sushi", "Burritos"}).SaveX(adminCtx)

Reviewer: ```suggestion
	hubUser := client.User.Create().SetName("a8m").SetTenant(hub).SaveX(adminCtx)
	labUser := client.User.Create().SetName("nati").SetTenant(lab).SaveX(adminCtx)
```

Please use these names instead if you find the previous example unclear. Also, the `foods` field is irrelevant for this example.

```

We finish our example with another privacy-rule named `DenyMismatchedTenants` on the `Group` schema.
The `DenyMismatchedTenants` rule rejects group creation if the associated users don't belong to
the same tenant as the group.

> ```go title="examples/privacytenant/rule/rule.go"
> // DenyMismatchedTenants is a rule that returns a deny decision if the operation
> // involves users that don't belong to the same tenant as the group or users that are from more than one tenant.

Reviewer: ```suggestion
```go title="examples/privacytenant/rule/rule.go"
// DenyMismatchedTenants is a rule that runs only on create operations and returns a deny
// decision if the operation tries to add users to groups that are not in the same tenant.
```

The `DenyMismatchedTenants` rule rejects the group creation if the associated users don't belong to
the same tenant as the group.

```go title="examples/privacytenant/rule/rule.go"
// DenyMismatchedTenants is a rule that returns a deny decision if the operation

> // involves users that don't belong to the same tenant as the group or users from more than one tenant.

Reviewer: Please mention that this rule is executed only on `Create` operations. See this line:

```go
privacy.OnMutationOperation(rule, ent.OpCreate)
```

Author: I made sure it is mentioned in the docs + in the code.

> // involves users that don't belong to the same tenant as the group or users are from more than one tenant.

Reviewer 2: ```suggestion
// involves users that don't belong to the same tenant as the group or users that are from more than one tenant.
```

		}
		// Query the tenant-id of all users. Expect to have exact 1 result,
		// and it matches the tenant-id of the group above.

> 		userTid, err := m.Client().User.Query().Where(user.IDIn(users...)).QueryTenant().OnlyID(ctx)

Reviewer: ```suggestion
		id, err := m.Client().User.Query().Where(user.IDIn(users...)).QueryTenant().OnlyID(ctx)
```
`id` is fine for a 3-lines scope.

		if err != nil {
			return privacy.Denyf("querying the tenant-id %v", err)
		}
		if userTid != tid {
			return privacy.Denyf("mismatch tenant-ids for group/users %d != %d", tid, userTid)
		}
		// Skip to the next privacy rule (equivalent to return nil).
		return privacy.Skip
	})
}
```

We add this rule to the `Group` schema and run code-generation.

```go title="examples/privacytenant/ent/schema/group.go"
// Policy defines the privacy policy of the Group.
func (Group) Policy() ent.Policy {
	return privacy.Policy{
		Mutation: privacy.MutationPolicy{
			privacy.OnMutationOperation(
				rule.DenyMismatchedTenants(),

> 				ent.OpCreate),

Reviewer: tbh, I prefer the previous version. Can you please explain why do you think it's needed?

		},
	}
}
```

Again, we expect the privacy-rules to take effect on the client operations.

```go title="examples/privacytenant/example_test.go"
func Do(ctx context.Context, client *ent.Client) error {
    // A continuation of the code-block above.

	// Expect operation to fail, because the DenyMismatchedTenants rule makes sure
	// the group and the users are connected to the same tenant.
	if _, err = client.Group.Create().SetName("entgo.io").SetTenant(hub).AddUsers(natiLab).Save(adminCtx); !errors.Is(err, privacy.Deny) {
		return fmt.Errorf("expect operation to fail, since user (natiLab) is not connected to the same tenant")
	}
	if _, err = client.Group.Create().SetName("entgo.io").SetTenant(hub).AddUsers(natiLab, a8mHub).Save(adminCtx); !errors.Is(err, privacy.Deny) {
		return fmt.Errorf("expect operation to fail, since some users (natiLab) are not connected to the same tenant")
	}

> 	entgo, err := client.Group.Create().SetName("entgo.io").SetTenant(hub).AddUsers(a8mHub).Save(adminCtx)

Reviewer: Please revert to the previous version, or explain why this one is preferred.

These lines just feel too long for me.

		}
		// Query the tenant-id of all users. Expect to have exact 1 result,
		// and it matches the tenant-id of the group above.
		userTid, err := m.Client().User.Query().Where(user.IDIn(users...)).QueryTenant().OnlyID(ctx)
		if err != nil {

> 			return privacy.Denyf("querying the tenant-id %v", err)

Reviewer 2: Are we sure we want `%v` here? I'm not sure of the internals, but ideally the error is embedded with `%w` so it can be unwrapped.

Author: You are correct, we do want `%w`. This is a mistake from a previous PR - #1275 it changed the `.md` file but not the example file `rule.go`, so when I copied it, it was overwritten. Fixing it.

		}
		// Query the tenant-id of all users. Expect to have exact 1 result,
		// and it matches the tenant-id of the group above.
		id, err := m.Client().User.Query().Where(user.IDIn(users...)).QueryTenant().OnlyID(ctx)
		if err != nil {

> 			return privacy.Denyf("querying the tenant-id %v", err)

Reviewer: ```suggestion
			return privacy.Denyf("querying the tenant-id %w", err)
```

		}
		if id != tid {
			return privacy.Denyf("mismatch tenant-ids for group/users %d != %d", tid, id)
		}
		// Skip to the next privacy rule (equivalent to return nil).
		return privacy.Skip
	})
}
```

We add this rule to the `Group` schema and run code-generation.

```go title="examples/privacytenant/ent/schema/group.go"
// Policy defines the privacy policy of the Group.
func (Group) Policy() ent.Policy {
	return privacy.Policy{
		Mutation: privacy.MutationPolicy{
			// Limit DenyMismatchedTenants only for
			// Create operation
			privacy.OnMutationOperation(
				rule.DenyMismatchedTenants(),
				ent.OpCreate,
			),
		},
	}
}
```

Again, we expect the privacy-rules to take effect on the client operations.

```go title="examples/privacytenant/example_test.go"
func Do(ctx context.Context, client *ent.Client) error {
    // A continuation of the code-block above.

> 	// Expect operation to fail, because the DenyMismatchedTenants rule makes sure
> 	// the group and the users are connected to the same tenant.

Reviewer: ```suggestion
	// Expect operation to fail because the DenyMismatchedTenants rule
	// makes sure the group and the users are connected to the same tenant.
```

```

In some cases, we want to reject user operations on entities that don't belong to their tenant **without loading

> these entities from the database** (unlike the `DenyMismatchedTenants` example above).  
> To achieve this, we can use the `FilterTenantRule` rule for mutations as well, but limit it to specific operations as follows:

Reviewer: ```suggestion
these entities from the database** (unlike the `DenyMismatchedTenants` example above). 
To achieve this, we can use the `FilterTenantRule` rule for mutations as well, but limit it to specific operations as follows:
```

```go title="examples/privacytenant/ent/schema/group.go"
// Policy defines the privacy policy of the Group.
func (Group) Policy() ent.Policy {
	return privacy.Policy{
		Mutation: privacy.MutationPolicy{

> 			privacy.OnMutationOperation(
> 				rule.DenyMismatchedTenants(),
> 				ent.OpCreate),

Reviewer: ```suggestion
			privacy.OnMutationOperation(
				rule.DenyMismatchedTenants(),
				ent.OpCreate,
			),
```

Then, we expect the privacy-rules to take effect on the client operations.

```go title="examples/privacytenant/example_test.go"
func Do(ctx context.Context, client *ent.Client) error {
    // A continuation of the code-block above.

> 	// Expect operation to fail, because the FilterTenantRule rule makes sure
> 	// that tenants can update and delete only their groups.
> 	if err = entgo.Update().SetName("fail.go").Exec(labView); !ent.IsNotFound(err) {
> 		return fmt.Errorf("expect operation to fail, since the group (entgo) is managed by a different tenant (hub), but got %w", err)
> 	}
> 	entgo, err = entgo.Update().SetName("entgo").Save(hubView)
> 	if err != nil {
> 		return fmt.Errorf("expect operation to pass, but got %w", err)
> 	}
> 	fmt.Println(entgo)
> 
> 	return nil

Reviewer: ```suggestion
	err = entgo.Update().SetName("fail.go").Exec(labView)
	// Expect operation to fail, because the FilterTenantRule rule makes
	// sure that tenants can update and delete only their groups.
	if !ent.IsNotFound(err) {
		return fmt.Errorf("expect operation to fail, since the group (entgo) is managed by a different tenant (hub), but got %w", err)
	}
	entgo, err = entgo.Update().SetName("entgo").Save(hubView)
	if err != nil {
		return fmt.Errorf("expect operation to pass, but got %w", err)
	}
	fmt.Println(entgo)
	return nil
```

Then, we expect the privacy-rules to take effect on the client operations.

```go title="examples/privacytenant/example_test.go"
func Do(ctx context.Context, client *ent.Client) error {
    // A continuation of the code-block above.

	// Expect operation to fail, because the FilterTenantRule rule makes sure
	// that tenants can update and delete only their groups.

> 	err = entgo.Update().SetName("fail.go").Exec(labView)
> 	if !ent.IsNotFound(err) {
> 		return fmt.Errorf("expect operation to fail, since the group (entgo) is managed by a different tenant (hub), but got %w", err)
> 	}

Reviewer 2: nit: limit error scope
```suggestion
	if err := entgo.Update().SetName("fail.go").Exec(labView); !ent.IsNotFound(err) {
		return fmt.Errorf("expect operation to fail, since the group (entgo) is managed by a different tenant (hub), but got %w", err)
	}
```

Then, we expect the privacy-rules to take effect on the client operations.

```go title="examples/privacytenant/example_test.go"
func Do(ctx context.Context, client *ent.Client) error {
    // A continuation of the code-block above.

	// Expect operation to fail, because the FilterTenantRule rule makes sure
	// that tenants can update and delete only their groups.
	err = entgo.Update().SetName("fail.go").Exec(labView)
	if !ent.IsNotFound(err) {
		return fmt.Errorf("expect operation to fail, since the group (entgo) is managed by a different tenant (hub), but got %w", err)
	}
	entgo, err = entgo.Update().SetName("entgo").Save(hubView)
	if err != nil {
		return fmt.Errorf("expect operation to pass, but got %w", err)
	}
	fmt.Println(entgo)

> 

Reviewer: ```suggestion
```
